File: 525.contiguous-array.cpp
/*
 * @lc app=leetcode id=525 lang=cpp
 *
 * [525] Contiguous Array
 */
#include <bits/stdc++.h>

using std::vector;

// @lc code=start
class Solution {
public:
    // Prefix sum + two sum
    // Treat all 0 as -1, then it becomes a special case of Q325 with k = 0
    // T: O(n)
    // S: O(n)
    int findMaxLength(vector<int>& nums) {
        std::unordered_map<int, int> mp;  // k: sum, v: index
        int sum = 0, ans = 0;

        mp[0] = -1;  // IMPORTANT: handle edge case if the subarr starts from idx 0
        for (int i = 0; i < nums.size(); ++i) {
            sum += nums[i] == 1 ? 1 : -1;
            auto it = mp.find(sum);
            if (it != mp.end()) {
                // Same prefix sum found, which means the sum of 
                // the numbers in the interval in between is 0
                ans = std::max(ans, i - it->second);
            } else {
                mp[sum] = i;
            }
        }

        return ans;
    }
};
// @lc code=end

File: 525.contiguous-array.py
#
# @lc app=leetcode id=525 lang=python3
#
# [525] Contiguous Array
#
from typing import List

# @lc code=start
class Solution:
    # T: O(n)
    # S: O(n)
    def findMaxLength(self, nums: List[int]) -> int:
        prefix_sum_idxs = {
            0: -1  # handle edge case like if the subarray begins from index 0
        }

        prefix_sum = 0
        ans = 0

        for i in range(len(nums)):
            # The target subarray sum should equals 0
            prefix_sum += 1 if nums[i] == 1 else -1

            last_idx = prefix_sum_idxs.get(prefix_sum, None)
            if last_idx is not None:
                ans = max(ans, i - last_idx)
            else:
                prefix_sum_idxs[prefix_sum] = i

        return ans

# @lc code=end

